Plutarch/DataRepr/Internal.hs

```diff
@@ -180,7 +186,161 @@ type family PUnLabel (a :: PLabeledType) :: PType where
+    
+    a) 'IndexedSum'. This is basically a singleton list a la SList from sop-core, except it can only represent nonempty lists, has an 
+       "explicit" inductive structure, and indicates the position of elements explicitly. 
+
+    b) A 'Length' type family. This doesn't work like 'length' from Prelude: 'Length '[a]' == 0, not 1. 
+
+    c) A type class 'IndexableSum' which is like SListI from sop-core, except for our 'IndexedSum' type. As far as I can tell it is impossible 
+       to reconstruct the sum without an inductive GADT that represents the structure of the sum. We need to conjure that structure out of thin air,
+       so we need a type class. 
+
+    d) Some dictionary stuff and an axiom to placate GHC and make this all typecheck.
+-}
+
+data PDataSum' :: Natural -> Bool -> [[PLabeledType]] -> PType where 
+  SumZ_ :: forall {x:: [PLabeledType]} {s :: S}
+           . PDataSum' 0 'False '[x] s
+
+  SumZ :: forall {x:: [PLabeledType]} {s :: S}
+           . (forall {out :: PType}. (Term s (PDataRecord x) -> Term s out) -> Term s out) 
+           -> PDataSum' 0 'True '[x] s
+
+  Inj :: forall {n :: Natural} {x :: [PLabeledType]} {xss :: [[PLabeledType]]} {s :: S} 
+        . (KnownNat (n+1))
+        => (forall {out :: PType}. (Term s (PDataRecord x) -> Term s out) -> Term s out) 
+        -> PDataSum' n 'False xss s 
+        -> PDataSum' (n + 1) 'True (x ': xss) s 
+
+  Inj_ :: forall {n :: Natural} {b :: Bool} {x :: [PLabeledType]} {xss :: [[PLabeledType]]} {s :: S} 
+        .  KnownNat (n+1)
+        => PDataSum' n b xss s 
+        -> PDataSum' (n + 1) b (x ': xss) s
+
+-- Used to reconstruct the PDataSum in the pmatch' instance for PlutusType 
+data IndexedSum :: Nat -> [k] -> Type where 
+  IxZ :: forall k (a :: k). IndexedSum 0 '[a]
+
+  IxS :: forall k (a :: k) (as :: [k]) 
+       . IndexedSum (Length as) as -> IndexedSum (Length as + 1) (a ': as)
+
+-- Well, length minus one at least 
+type family Length (xss :: [k]) :: Nat where 
+  Length '[x] = 0 
+  Length (x ': xs) = (Length xs) + 1 
+
+-- Could convert an SList into this, but this is simpler (I don't think SList is suitable here)
+class IndexableSum (defs :: [k]) where 
+  indexSum :: IndexedSum (Length defs) defs 
+
+-- this is kind of a cheap hack and there's probably a better way 
+knownLength :: forall n ks. IndexedSum n ks -> Dict (KnownNat n)
+knownLength = \case 
+  IxZ -> Dict 
+  ixed@(IxS rest) -> case knownLength rest of 
+    d@Dict -> go ixed d 
+ where 
+   go :: forall (n' :: Nat) a as . IndexedSum (n' + 1) (a:as) 
+      -> Dict (KnownNat n') 
+      -> Dict (KnownNat (n' + 1))
+   go (IxS _) Dict = mapDict plusNat (Dict :: Dict (KnownNat n', KnownNat 1))
+
+-- maybe don't need the constraints? 
+addLength :: forall k ks. (IndexableSum (k:ks), IndexableSum ks) => Length (k:ks) :~: Length ks + 1 
+addLength = UNSAFE.unsafeCoerce Refl 
```

Comment:
this is really bad....

Reply:
Ugh, yeah. I'll find some way to get rid of it.